File: 735-asteroid-collision/solution.py
class Solution:
    def getTop(self, stack: List[int]) -> int:
        top = stack[-1] if stack else None
        return top

    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        a = []
        for idx, i in enumerate(asteroids):
            while True:
                top = self.getTop(a)
                if i > 0 or top == None or top*i > 0:
                    a.append(i)
                    break
                if abs(top) > abs(i):
                    # if top is gt, it wins, so dont add i to stack
                    break
                elif abs(top) == abs(i):
                    a.pop()
                    # if top == i, its a draw. a is (--) from the stack, and i never gets (++) to stack top
                    break
                else:
                    # if 
                    a.pop()
        return a

File: 735-asteroid-collision/solution.go
package main

import (
	"math"
)


type stack []int

func (s stack) Top() *int {
    if len(s) == 0 {
        return nil
    }
    return &s[len(s)-1]
}

func (s stack) Pop() (stack, *int) {
	if len(s) == 0 {
        return s, nil
    }
    return s[:len(s)-1], &s[len(s)-1]
}

func (s stack) Push(i int) stack {
    return append(s, i)
}

func asteroidCollision(asteroids []int) []int {
	var s stack
    for _, i := range asteroids {
        for {
			top := s.Top()
			if top == nil {
				s = s.Push(i)
				break
			}

			if *top*i > 0 || i > 0 {
				s = s.Push(i)
				break
			} else {
				if math.Abs(float64(*top)) > math.Abs(float64(i)) {
					break
				} else if math.Abs(float64(*top)) == math.Abs(float64(i)) {
					s, _ = s.Pop()
					break
				} else {
					s, _ = s.Pop()
				}
			}
		}
    }
	return s
}